### ISSUE TITLE
NIOLock and NIOLockedValueBox aren't strict enough

### ISSUE THREAD

**Issue by Lukasa** [apple/swift-nio#2910]
NIOLock is particularly scary, but NIOLockedValueBox has some limitations too that need to be tidied up.

The following functions are dangerous:

https://github.com/apple/swift-nio/blob/9ff5fddb3d4da3b62dadb8ff83944acf6b0eabf9/Sources/NIOConcurrencyHelpers/NIOLock.swift#L232-L254

Here, `withLock` needs to take its closure `sending` and the return value needs to be `sending` too. In Swift 5, these both need to be `Sendable`, with an unchecked version available.

https://github.com/apple/swift-nio/blob/9ff5fddb3d4da3b62dadb8ff83944acf6b0eabf9/Sources/NIOConcurrencyHelpers/NIOLockedValueBox.swift#L35-L39

Here, `withLockedValue` needs again to take its closure `sending`. Optionally, we could relax its `Sendable` requirement, and also return the value `sending`, but given our need to remain compatible with Swift 5 we should probably leave it stricter for now. In Swift 5, the closure probably needs to be `@Sendable`.

**Comment by FranzBusch** [comment]
Yep good catch. We need to basically copy the API from `Mutex` on the `LockedValueBox` here

**Comment by weissi** [comment]
I think we should reverb this ticket as it's currently written confusingly. My summary would be:

- Most importantly: `NIOLockedValueBox` as well as `NIOLock` are correct regarding sendability etc
- `NIOLock` should stay is as and should _not_ get a `sending` in `withLock` (as that would make it less useful)
- `NIOLockedValueBox` is _more_ strict than it needs to be as of Swift 6.0. It could be even more helpful if it took `sending` closures as that would allow us to make it unconditionally `Sendable`.

**Comment by weissi** [comment]
@Lukasa can we close this bug or retitle to "NIOLockedValueBox _too strict_"?
